framework/src/bcs/IntegratedBC.C

```diff
@@ -27,65 +27,28 @@ template <>
 InputParameters
 validParams<IntegratedBC>()
 {
-  InputParameters params = validParams<BoundaryCondition>();
-  params += validParams<RandomInterface>();
-  params += validParams<MaterialPropertyInterface>();
-
-  params.addParam<std::vector<AuxVariableName>>(
-      "save_in",
-      "The name of auxiliary variables to save this BC's residual contributions to.  "
-      "Everything about that variable must match everything about this variable (the "
-      "type, what blocks it's on, etc.)");
-  params.addParam<std::vector<AuxVariableName>>(
-      "diag_save_in",
-      "The name of auxiliary variables to save this BC's diagonal jacobian "
-      "contributions to.  Everything about that variable must match everything "
-      "about this variable (the type, what blocks it's on, etc.)");
-
-  params.addParamNamesToGroup("diag_save_in save_in", "Advanced");
-
-  // Integrated BCs always rely on Boundary MaterialData
-  params.set<Moose::MaterialDataType>("_material_data_type") = Moose::BOUNDARY_MATERIAL_DATA;
-
+  InputParameters params = validParams<IntegratedBCBase>();
   return params;
 }
 
 IntegratedBC::IntegratedBC(const InputParameters & parameters)
-  : BoundaryCondition(parameters, false), // False is because this is NOT nodal
-    RandomInterface(parameters, _fe_problem, _tid, false),
-    CoupleableMooseVariableDependencyIntermediateInterface(this, false),
-    MaterialPropertyInterface(this, boundaryIDs()),
-    _current_elem(_assembly.elem()),
-    _current_elem_volume(_assembly.elemVolume()),
-    _current_side(_assembly.side()),
-    _current_side_elem(_assembly.sideElem()),
-    _current_side_volume(_assembly.sideElemVolume()),
-
+  : IntegratedBCBase(parameters), // False is because this is NOT nodal
+    _var(dynamic_cast<MooseVariable &>(_base_var)),
```

Comment:
@andrsd Would templating the getVariable methods make sense?

Reply:
Yes, it would.

Reply:
Although you might have an issue in `AuxiliarySystem`, because your PR does not allow vector-valued aux-variables. But, you should be able to use explicit instantiation to prevent people to add/get them.

Reply:
What design do you like better: 
1) A single container of MooseVariableBase pointers in the VariableWarehouse with `dynamic_cast` checking in the templated `SystemBase` getter
2) Separate (unordered)maps of MooseVariables and MooseVariableVectors and a templated getter method in the VariableWarehouse that removes the need for dynamic casting in the `SystemBase` getter

Reply:
I'd go with 2. We already have normal + scalar variables there, so it is just logical we add vector valued variables there.